src/main/java/io/projectriff/invoker/FunctionConfiguration.java

```diff
@@ -265,18 +269,42 @@ public Object create(String type) {
 					.overrideThreadContextClassLoader(functionClassLoader);
 			AutowireCapableBeanFactory factory = context.getAutowireCapableBeanFactory();
 			try {
+				Object result = null;
 				if (this.runner != null) {
-					return this.runner.getBean(type);
+					result = this.runner.getBean(type);
+				}
+				if (result == null) {
+					logger.info("No main class provided. Instantiating: " + type);
+					if (ClassUtils.isPresent(type, functionClassLoader)) {
+						result = factory.createBean(
+								ClassUtils.resolveClassName(type, functionClassLoader));
+					}
+				}
+				if (result != null) {
+					return result;
 				}
-				logger.info("No main class provided. Instantiating: " + type);
-				return factory.createBean(
-						ClassUtils.resolveClassName(type, functionClassLoader));
+				throw new IllegalStateException("Cannot create bean for: " + type);
 			}
 			finally {
 				ClassUtils.overrideThreadContextClassLoader(contextClassLoader);
 			}
 		}
 
+		private Object function(FunctionRegistration<Object> reg) {
```

Comment:
This function seems to encapsulate some really important design decisions, although I can't easily reverse engineer those from the code. Would it be possible to add some tests as a way of documenting what behaviour is expected? I think it would benefit from some comments too.

Reply:
There are two integration tests (see changes to tests below). Comments coming...